src/jdk.jdwp.agent/share/native/libjdwp/invoker.c

```diff
@@ -232,6 +232,10 @@ fillInvokeRequest(JNIEnv *env, InvokeRequest *request,
     /*
      * Squirrel away the method signature
      */
+    char* oldSignature = request->methodSignature;
+    if (oldSignature != NULL) {
+      jvmtiDeallocate(oldSignature);
+    }
     error = methodSignature(method, NULL, &request->methodSignature,  NULL);
     if (error != JVMTI_ERROR_NONE) {
```

Comment:
Shouldn't this be done when the invoke has completed rather than at the start of the next invoke? Otherwise you potentially have one outstanding allocation for every thread, and you still have a leak when the thread exits.

Reply:
> Shouldn't this be done when the invoke has completed rather than at the start of the next invoke? Otherwise you potentially have one outstanding allocation for every thread, and you still have a leak when the thread exits.

Yes, perhaps. Please help me, where would be a good and reliable place to do that? Towards the end of invoker_doInvoke() ?

Reply:
In `invoker_completeInvokeRequest()` this appears to be the last reference:

`        jbyte returnType = methodSignature_returnTag(request->methodSignature);`

I would suggest freeing outside of the `if (!detached)` block and setting it to `NULL`. You might want to add an assert for `NULL` where you are currently freeing the pointer.

Reply:
> Are you still hitting the assert in fillInvokeRequest()? I'm not sure why it would ever not be NULL there. It's probably worth investigating some more. Otherwise you need to remove the assert, and possibly still have a leak.

No, the latest version seems good and passes tier1-3 and customer confirmed that it fixes the memory leak.

Reply:
Ok. I think I prefer you to free the memory after calling `deleteGlobalArgumentRefs()` since `deleteGlobalArgumentRefs()` is really all about releasing JNI global refs, not freeing other allocated memory.

Reply:
Ok, I've moved it back to after the call to deleteGlobalArgumentRefs().